crates/transport/src/layers/fallback.rs

```diff
@@ -0,0 +1,453 @@
+use std::{
+    collections::VecDeque,
+    fmt::Debug,
+    num::NonZeroUsize,
+    sync::Arc,
+    task::{Context, Poll},
+    time::{Duration, Instant},
+};
+
+use alloy_json_rpc::{RequestPacket, ResponsePacket};
+use derive_more::{Deref, DerefMut};
+use futures::{stream::FuturesUnordered, StreamExt};
+use parking_lot::RwLock;
+use tower::{Layer, Service};
+use tracing::{debug, trace};
+
+use crate::{TransportError, TransportErrorKind, TransportFut};
+
+// Constants for the transport ranking algorithm
+const STABILITY_WEIGHT: f64 = 0.7;
+const LATENCY_WEIGHT: f64 = 0.3;
+const DEFAULT_SAMPLE_COUNT: usize = 10;
+const DEFAULT_ACTIVE_TRANSPORT_COUNT: usize = 3;
+
+/// The [`FallbackService`] consumes multiple transports and is able to
+/// query them in parallel, returning the first successful response.
+///
+/// The service ranks transports based on latency and stability metrics,
+/// and will attempt to always use the best available transports.
+#[derive(Debug, Clone)]
+pub struct FallbackService<S> {
+    /// The list of transports to use
+    transports: Arc<RwLock<Vec<ScoredTransport<S>>>>,
+    /// The maximum number of transports to use in parallel
+    active_transport_count: usize,
+    /// Whether to log transport rankings
+    log_transport_rankings: bool,
+}
+
+impl<S> FallbackService<S> {
+    /// Create a new fallback service from a list of transports.
+    ///
+    /// - The `active_transport_count` parameter controls how many transports are used for requests
+    ///   at any one time.
+    /// - The `log_transport_rankings` parameter controls whether the current transport rankings are
+    ///   output as traces. In order to see them, make sure to enable the `alloy_transport` target
+    ///   (i.e. `RUST_LOG=alloy_transport=debug`).
+    pub fn new(
+        transports: Vec<S>,
+        active_transport_count: usize,
+        log_transport_rankings: bool,
```

Comment:
can we make this bool a separate setter instead, but I'm also okay with always emitting this as trace!